src/Nethermind/Nethermind.TxPool/Collections/SortedPool.cs

```diff
@@ -18,102 +18,99 @@
+    public class SortedPool<TKey, TValue, TGroup>
     {
         private readonly int _capacity;
-        protected readonly Comparison<TValue> Comparison;
-        protected readonly Dictionary<TKey, LinkedListNode<KeyValuePair<TKey, TValue>>> CacheMap;
-        protected readonly LinkedList<KeyValuePair<TKey, TValue>> LruList;
-
-        public SortedPool(int capacity, Comparison<TValue> comparison)
+        private readonly IComparer<TValue> _comparerWithIdentity;
+        private readonly Func<TValue, TGroup> _groupMapping;
+        private readonly IDictionary<TGroup, ICollection<TValue>> _buckets;
+        private readonly DictionarySortedSet<TValue, TKey> _sortedValues;
+        private readonly IDictionary<TKey, TValue> _cacheMap;
+        
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="capacity">Max capacity</param>
+        /// <param name="comparerWithIdentity">Comparer to sort items. It must differentiate items by their identity or some items will be lost.</param>
+        /// <param name="groupMapping">Mapping from <see cref="TValue"/> to <see cref="TGroup"/></param>
+        public SortedPool(int capacity, IComparer<TValue> comparerWithIdentity, Func<TValue, TGroup> groupMapping)
         {
             _capacity = capacity;
-            Comparison = comparison;
-            CacheMap = new Dictionary<TKey, LinkedListNode<KeyValuePair<TKey, TValue>>>(); // do not initialize it at the full capacity
-            LruList = new LinkedList<KeyValuePair<TKey, TValue>>();
+            _comparerWithIdentity = comparerWithIdentity ?? throw new ArgumentNullException(nameof(comparerWithIdentity));
+            _groupMapping = groupMapping ?? throw new ArgumentNullException(nameof(groupMapping));
+            _cacheMap = new Dictionary<TKey, TValue>(); // do not initialize it at the full capacity
+            _buckets = new Dictionary<TGroup, ICollection<TValue>>();
+            _sortedValues = new DictionarySortedSet<TValue, TKey>(_comparerWithIdentity);
         }
 
-        public int Count => CacheMap.Count;
+        public int Count => _cacheMap.Count;
 
         [MethodImpl(MethodImplOptions.Synchronized)]
         public TValue[] GetSnapshot()
         {
-            return LruList.Select(i => i.Value).ToArray();
+            return _buckets.SelectMany(b => b.Value).ToArray();
         }
-
+        
         [MethodImpl(MethodImplOptions.Synchronized)]
-        public TValue TakeFirst()
+        public IDictionary<TGroup, TValue[]> GetBucketSnapshot()
         {
-            var value = LruList.First.Value;
-            LruList.RemoveFirst();
-            Remove(value.Key);
-            return value.Value;
+            return _buckets.ToDictionary(g => g.Key, g => g.Value.ToArray());
         }
         
         [MethodImpl(MethodImplOptions.Synchronized)]
-        public bool TryRemove(TKey key, out TValue tx)
+        public TValue TakeFirst()
+        {
```

Comment:
If this was named TryTakeFirst then it can return false on empty collection. Now it returns null which is not a typical API